Sources/Testing/Polling/Polling.swift

```diff
@@ -0,0 +1,463 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+/// Default values for polling confirmations.
+@available(_clockAPI, *)
+internal let defaultPollingConfiguration = (
+  pollingDuration: Duration.seconds(1),
+  pollingInterval: Duration.milliseconds(1)
+)
+
+/// A type defining when to stop polling.
+/// This also determines what happens if the duration elapses during polling.
+@_spi(Experimental)
+public enum PollingStopCondition: Sendable, Equatable, Codable {
+  /// Evaluates the expression until the first time it returns true.
+  /// If it does not pass once by the time the timeout is reached, then a
+  /// failure will be reported.
+  case firstPass
+
+  /// Evaluates the expression until the first time it returns false.
+  /// If the expression returns false, then a failure will be reported.
+  /// If the expression only returns true before the timeout is reached, then
+  /// no failure will be reported.
+  /// If the expression does not finish evaluating before the timeout is
+  /// reached, then a failure will be reported.
+  case stopsPassing
+}
+
+/// A type describing why polling failed
+@_spi(Experimental)
+public enum PollingFailureReason: Sendable, Codable {
```

Comment:
I think this `PollingFailureReason` enum would feel more natural nested within `PollingFailedError`, and then its name could be simplified to `Reason` (or fully-qualified, `PollingFailedError.Reason`).

Reply:
Fair idea. The main reason for keeping `PollingFailedError` a struct is to make the conformance to `CustomIssueRepresentable` simple and obvious.

For what test authors consume, I expect `PollingFailedError.Reason` provides enough information for them to do anything custom - they can't really access `sourceContext` in the general case, and `comment` likely isn't that useful to handle?

I'm inclined to keep with @stmontgomery's suggestion right now, but I definitely see @grynspan's argument.